detox/src/devices/drivers/SimulatorDriver.js

```diff
@@ -157,6 +155,73 @@ class SimulatorDriver extends IosDriver {
   async waitForBackground() {
     return await this.client.waitForBackground();
   }
+
+  /***
+   * @private
+   * @param deviceQuery
+   * @returns {Promise<String>}
+   */
+  async _findOrCreateDevice(deviceQuery) {
+    let udid;
+
+    const { free, busy } = await this._groupDevicesByStatus(deviceQuery);
+
+    if (_.isEmpty(free)) {
+      const prototypeDevice = busy[0];
+      udid = this.applesimutils.create(prototypeDevice);
+    } else {
+      udid = free[0].udid;
+    }
+
+    if (!udid) {
+      throw new Error(`Failed to find device matching "${deviceQuery}"`);
+    }
+
+    return udid;
+  }
+
+  async _groupDevicesByStatus(query) {
```

Comment:
Query should be an object rather than string. If we are making such important changes, let's do it fully. Otherwise, there is ambiguity that cannot be resolved with this query. Devices with the same name are allowed to exist, If I have two devices named "MySim", one for iOS 13 and one for iOS 12, the current proposed solution will fail for a `"MySim, iOS 13.0"` query.

Reply:
Perhaps we can have a fallback like this for legacy strings, but we should add the full solution. I don't think it's a hard solution, as the object can travel, just like the string, from the package.json to here.

Reply:
I decided to split the refactor and the change to the feature, on Friday it will appear in a consequent pull request.